src/UserSpaceInstrumentation/AccessTraceesMemory.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "AccessTraceesMemory.h"
+
+#include <absl/strings/numbers.h>
+#include <absl/strings/str_format.h>
+#include <absl/strings/str_split.h>
+#include <sys/ptrace.h>
+
+#include <cerrno>
+#include <cstring>
+#include <string>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/ReadFileToString.h"
+#include "OrbitBase/SafeStrerror.h"
+
+namespace orbit_user_space_instrumentation {
+
+using orbit_base::ReadFileToString;
+
+[[nodiscard]] ErrorMessageOr<void> ReadTraceesMemory(pid_t pid, uint64_t address_start,
+                                                     uint64_t length, std::vector<uint8_t>* bytes) {
+  // Round up length to next multiple of eight.
+  length = ((length + 7) / 8) * 8;
+  bytes->resize(length);
+  for (size_t i = 0; i < length / 8; i++) {
+    const uint64_t data = ptrace(PTRACE_PEEKDATA, pid, address_start + i * 8, NULL);
+    if (errno) {
+      return ErrorMessage(absl::StrFormat("Failed to PTRACE_PEEKDATA with errno %d: \"%s\"", errno,
+                                          SafeStrerror(errno)));
+    }
+    std::memcpy(bytes->data() + (i * sizeof(uint64_t)), &data, 8);
+  }
+  return outcome::success();
+}
+
+[[nodiscard]] ErrorMessageOr<void> WriteTraceesMemory(pid_t pid, uint64_t address_start,
+                                                      const std::vector<uint8_t>& bytes) {
+  size_t pos = 0;
+  do {
+    // Pack 8 byte for writing into `data`.
+    uint64_t data = 0;
+    std::memcpy(&data, bytes.data() + pos, sizeof(uint64_t));
+    if (ptrace(PTRACE_POKEDATA, pid, address_start + pos, data) == -1) {
+      ErrorMessage(absl::StrFormat("Unable to write data into tracees (pid: %d) memory.", pid));
```

Comment:
OCD thought: This message is very different from the one in `ReadTraceesMemory`, but the two functions have a clear parallel. I would make them similar if it's possible.

Reply:
Changed it a bit. The functions are more different than it might appear though: PTRACE_POKEDATA returns -1 on error. PTRACE_PEEKDATA sets errno.